Port remaining benchmarks to BenchmarkDotNet

adamsitnik opened the pull request.

adamsitnik commented:
@dotnet-bot test Innerloop Ubuntu16.04 Release Build and Test

ahsonkhan commented on tests/Benchmarks/Benchmarks.csproj:
```diff
@@ -9,6 +9,11 @@
     <NoWarn>$(NoWarn);CS0618</NoWarn>
   </PropertyGroup>
+  <ItemGroup>
+    <Compile Remove="BenchmarkDotNet.Artifacts\**" />
```
What is the purpose of this ItemGroup?

adamsitnik replied:
by default, VS displays the content of BenchmarkDotNet.Artifacts folder in the solution explorer. The folder contains results and logs. This changes is the result of right clicking on the folder and choosing "exclude from project"

ahsonkhan commented on tests/Benchmarks/Pipelines/E2EPipeline.cs:
```diff
@@ -0,0 +1,69 @@
+
+namespace Benchmarks.Pipelines
+{
+    [SimpleJob(RunStrategy.Monitoring)] // the setting for benchmarks with not-steady state
```
Why is this necessary here? We previously ran it multiple times and reported the result as part of xunit.perf.

ahsonkhan replied:
Any thoughts on this, @adamsitnik?

adamsitnik replied:
@ahsonkhan The default settings of BenchmarkDotNet use some heuristics to determine how many times to run the benchmark to get stable results. This is perfect for steady-state benchmarks. 

However, benchmarks which don't have a steady state (like heavy IO benchmarks) should not use the default settings. Their distribution is typically very different, which does not work well with our heuristics.

[Monitoring](https://benchmarkdotnet.org/Guides/ChoosingRunStrategy.htm) strategy runs it a certain amount of times, does not pre-warmup the code and does not exclude the outliers. It's preferred strategy for benchmarks like this one.

ahsonkhan commented on tests/Benchmarks/Program.cs:
```diff
@@ -17,9 +18,17 @@ public class Program
+        private static Type[] GetGenericBenchmarks()
+            => new[]
+            {
+                typeof(HttpParser<Request>),
```
How do we extend this when new tests are added with different generic types?

adamsitnik replied:
In that case, we will need to specify the generic arguments like in this example.
Why is that? When BDN finds a generic type (let's say `Benchmark<T>`) it does not know which `T` should be used. So this has to be specified in an explicit way.

I just created an issue for that in BDN to allow doing this by using attributes: https://github.com/dotnet/BenchmarkDotNet/issues/745

ahsonkhan commented on tests/Benchmarks/System.Memory/IndexOf.cs:
```diff
@@ -0,0 +1,35 @@
+        [Benchmark(OperationsPerInvoke = 6 * 4)] // avoid loop overhead, do manuall loop unrolling
+        [Arguments(0)] // we don't have more arguments here because the CPU cache gets warmed-up very soon and the results are the same for every $at
+        [Arguments(1000)]
+        public int SpanIndexOf(int at)
```
The previous test was measuring the IndexOf method and comparing the vectorized vs. non-vectorized, not the span indexer.

Also, do we have a way to programmatically set Vector.IsHardwareAccelerated to true/false? I know it can be done by setting an environment variable before running the test.

adamsitnik replied:
@ahsonkhan yes, it's possible. Could you tell me the name of env var which controlls that? I will add sample config then

ahsonkhan replied:
You can disable vectorization by setting `COMPlus_FeatureSIMD=0`

ahsonkhan commented on tests/Benchmarks/System.Memory/Slice.cs:
```diff
@@ -0,0 +1,37 @@
+    /// A class containing benchmarks for comparing constructs that do the same task in
+    /// different ways with the same end outcome.
+    /// </summary>
+    // [HardwareCounters(HardwareCounter.InstructionRetired)] // uncomment if you want to get hardware counters
```
Can we un-comment this from the get go?

adamsitnik replied:
sure, but please remember that to get Hardware Counters you need to run benchmarks as Admin (it's ETW limitation, BDN just warns about it)

adamsitnik replied:
When running other benchmarks, it does not affect the experience.
When running this particular benchmark as non-admin, it's going to complain about it and run it without hardware counters.

So I think that we can keep it.

ahsonkhan commented on tests/Benchmarks/System.Memory/Slice.cs:
```diff
@@ -0,0 +1,37 @@
+﻿using BenchmarkDotNet.Attributes;
```
Missing license header

ahsonkhan commented on tests/Benchmarks/Http/HttpParser.cs:
```diff
@@ -32,38 +31,33 @@ public class HttpParser
     //public string Target;
 
-    public Dictionary<string, string> Headers = new Dictionary<string, string>();
+    //public Dictionary<string, string> Headers = new Dictionary<string, string>();
```
Was commenting this out necessary for some benchmark related reason or is it just cleanup to be consistent with the rest of the commented out code?

adamsitnik replied:
I commented it because the corresponding struct type has this field commented too. I wanted to make them more "equal" to make the comparison easier.

ahsonkhan commented on tests/Benchmarks/Azure/CosmosDb.cs:
```diff
@@ -35,26 +35,26 @@ public class CosmosDb
 
     [GlobalSetup]
-    public void CosmosDbBench()
+    public void Setup()
```
As part of this clean up change, can you add a System.Azure.Experimental.Benchmarks namespace and move this test to a System.Azure.Experimental directory under Benchmarks?

ahsonkhan commented on tests/Benchmarks/System.Text.Utf8/Utf8StringPerf.cs:
```diff
@@ -0,0 +1,49 @@
+﻿using System;
```
license header

ahsonkhan commented on tests/Benchmarks/System.Text.Utf8/Utf8StringPerf.cs:
```diff
@@ -0,0 +1,49 @@
+        [Benchmark]
+        public void EnumerateCodePoints()
+        {
+            foreach (var codePoint in utf8String)
```
nit: don't use var here

ahsonkhan commented on tests/Benchmarks/System.Text.Utf8/Utf8StringPerf.cs:
```diff
@@ -0,0 +1,49 @@
+using System.Text.Utf8;
+using BenchmarkDotNet.Attributes;
+
+namespace Benchmarks.System.Text.Utf8
```
We decided on the namespace structure (like unit tests) as `<namespace>.Benchmarks`

ahsonkhan commented on tests/Benchmarks/System.Text.Utf8/Utf8StringPerf.cs:
```diff
@@ -0,0 +1,49 @@
+        public int Length;
+
+        [Params(126, 0xD7FF)]
+        public int MaxCodePoint;
```
We end up losing the "description" of the input. "Short ASCII string", "Short string" etc. Any way we can get that back?

adamsitnik replied:
Unfortunately not ;(

adamsitnik commented:
@ahsonkhan I pushed the fixes:

1. Adopted all types with benchmarks to the naming convention
2. Added missing license header
3. Fixed the IndexOf benchmark and added the config which allows comparing SIMD vs no SIMD

ahsonkhan commented on tests/Benchmarks/System.Azure.Experimental/CosmosDb.cs:
```diff
@@ -0,0 +1,214 @@
+using System.Runtime.InteropServices;
+using BenchmarkDotNet.Attributes;
+
+
```
nit: extra new line

ahsonkhan commented on tests/Benchmarks/System.Memory/IndexOf.cs:
```diff
@@ -0,0 +1,60 @@
+        [GlobalSetup]
+        public void PrintInfo() => Console.WriteLine($"// Vector.IsHardwareAccelerated={Vector.IsHardwareAccelerated}"); // just print the info to make sure the config works
+
+        [Benchmark(OperationsPerInvoke = 6 * 4)] // avoid loop overhead, do manuall loop unrolling
```
Do we still need to do this and repeat the operation?

adamsitnik replied:
@ahsonkhan yes, because the cost of creating Span is a big part of the benchmark for the cases where the searched element is at the beginning of memory buffer. 

``` ini

BenchmarkDotNet=v0.10.14.525-nightly, OS=Windows 10.0.16299.309 (1709/FallCreatorsUpdate/Redstone3)
Intel Xeon CPU E5-1650 v4 3.60GHz, 1 CPU, 12 logical and 6 physical cores
Frequency=3507503 Hz, Resolution=285.1031 ns, Timer=TSC
.NET Core SDK=2.1.300-rc1-008662
  [Host]        : .NET Core 2.1.0-rc1-26424-03 (CoreCLR 4.6.26424.05, CoreFX 4.6.26424.04), 64bit RyuJIT
  SIMD Disabled : .NET Core 2.1.0-rc1-26424-03 (CoreCLR 4.6.26424.05, CoreFX 4.6.26424.04), 64bit RyuJIT
  SIMD Enabled  : .NET Core 2.1.0-rc1-26424-03 (CoreCLR 4.6.26424.05, CoreFX 4.6.26424.04), 64bit RyuJIT

LaunchCount=1  TargetCount=3  WarmupCount=3  

```
|      Method |           Job |  EnvironmentVariables |   at |       Mean |      Error |    StdDev |
|------------ |-------------- |---------------------- |----- |-----------:|-----------:|----------:|
| **SpanIndexOf** | **SIMD Disabled** | **COMPlus_FeatureSIMD=0** |    **0** |   **2.763 ns** |  **0.3031 ns** | **0.0171 ns** |
| SpanIndexOf |  SIMD Enabled | COMPlus_FeatureSIMD=1 |    0 |   5.815 ns |  2.5243 ns | 0.1426 ns |
| **SpanIndexOf** | **SIMD Disabled** | **COMPlus_FeatureSIMD=0** |    **1** |   **3.086 ns** |  **0.9012 ns** | **0.0509 ns** |
| SpanIndexOf |  SIMD Enabled | COMPlus_FeatureSIMD=1 |    1 |   5.997 ns |  6.9594 ns | 0.3932 ns |
| **SpanIndexOf** | **SIMD Disabled** | **COMPlus_FeatureSIMD=0** |    **2** |   **3.354 ns** |  **1.0120 ns** | **0.0572 ns** |
| SpanIndexOf |  SIMD Enabled | COMPlus_FeatureSIMD=1 |    2 |   5.776 ns |  2.7741 ns | 0.1567 ns |
| **SpanIndexOf** | **SIMD Disabled** | **COMPlus_FeatureSIMD=0** |    **4** |   **3.280 ns** |  **0.4101 ns** | **0.0232 ns** |
| SpanIndexOf |  SIMD Enabled | COMPlus_FeatureSIMD=1 |    4 |   5.710 ns |  2.1890 ns | 0.1237 ns |
| **SpanIndexOf** | **SIMD Disabled** | **COMPlus_FeatureSIMD=0** |    **8** |   **4.459 ns** |  **0.0775 ns** | **0.0044 ns** |
| SpanIndexOf |  SIMD Enabled | COMPlus_FeatureSIMD=1 |    8 |   5.941 ns |  0.8559 ns | 0.0484 ns |
| **SpanIndexOf** | **SIMD Disabled** | **COMPlus_FeatureSIMD=0** |   **16** |   **5.976 ns** |  **1.8004 ns** | **0.1017 ns** |
| SpanIndexOf |  SIMD Enabled | COMPlus_FeatureSIMD=1 |   16 |   6.020 ns |  3.3568 ns | 0.1897 ns |
| **SpanIndexOf** | **SIMD Disabled** | **COMPlus_FeatureSIMD=0** |   **30** |   **8.362 ns** |  **0.2031 ns** | **0.0115 ns** |
| SpanIndexOf |  SIMD Enabled | COMPlus_FeatureSIMD=1 |   30 |   6.405 ns |  3.3365 ns | 0.1885 ns |
| **SpanIndexOf** | **SIMD Disabled** | **COMPlus_FeatureSIMD=0** | **1000** | **225.725 ns** | **92.0383 ns** | **5.2003 ns** |
| SpanIndexOf |  SIMD Enabled | COMPlus_FeatureSIMD=1 | 1000 |  32.428 ns |  5.5044 ns | 0.3110 ns |
